aviatrix/resource_aviatrix_kubernetes_cluster.go

```diff
@@ -0,0 +1,291 @@
+package aviatrix
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"strings"
+
+	"github.com/hashicorp/go-cty/cty"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/AviatrixSystems/terraform-provider-aviatrix/v3/goaviatrix"
+)
+
+func resourceAviatrixKubernetesCluster() *schema.Resource {
+	return &schema.Resource{
+		CreateWithoutTimeout: resourceAviatrixKubernetesClusterCreate,
+		ReadWithoutTimeout:   resourceAviatrixKubernetesClusterRead,
+		UpdateWithoutTimeout: resourceAviatrixKubernetesClusterUpdate,
+		DeleteWithoutTimeout: resourceAviatrixKubernetesClusterDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				// will be computed if arn is set
+				Computed:     true,
+				Description:  "Id of the kubernetes cluster.",
+				ExactlyOneOf: []string{"cluster_id", "arn"},
+			},
+			"arn": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "AWS ARN of the cluster if it is an EKS cluster.",
+				ValidateDiagFunc: func(i interface{}, path cty.Path) diag.Diagnostics {
+					arn := i.(string)
+					if _, _, _, err := parseArn(arn); err != nil {
+						return diag.Errorf("invalid ARN: %s", err)
+					}
+					return nil
+				},
+			},
+			"use_csp_credentials": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Whether to use the credential of the account.",
+			},
+			"kube_config": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Kube config file content of the cluster.",
+				Sensitive:   true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					// The controller trims all encrypted content, so we need to trim the kube config content before comparing
+					oldTrimmed := strings.TrimSpace(old)
+					newTrimmed := strings.TrimSpace(new)
+					return oldTrimmed == newTrimmed
+				},
+			},
+			"account_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Name of the account with management privileges over the cluster",
+				RequiredWith: []string{"account_id", "name", "vpc_id", "region", "version", "platform", "is_publicly_accessible", "network_mode"},
```

Comment:
https://coda.io/d/Aviatrix-Terraform-Provider-Style-Guide_d_YQovtoErX/Nested-Configuration_su_Kr?utm_source=slack#_luJ6D 

For these fields that are all required with each other. I think it would be cleaner to have a nested object. With TF SDK v2 I think that means we would have a TypeList field with max 1 element.

Reply:
Sure, I can circle back to that.

Reply:
I updated the PR to expect the cluster configuration in a block `cluster_details`:
```hcl
resource "aviatrix_kubernetes_cluster" "mycluster" {
  cluster_id = "mycluster"

  kube_config = var.kubeconfig

  cluster_details {
    account_name           = data.aviatrix_account.aws.account_name
    account_id             = data.aviatrix_account.aws.aws_account_number
    name                   = "mycluster"
    region                 = "us-east-2"
    vpc_id                 = data.aws_vpc.vpc.id
    is_publicly_accessible = true
    platform               = "kops"
    version                = "1.30"
    network_mode           = "OVERLAY"
    tags = {
      foo = "bar"
    }
  }
}
```
This results in the following config in Mongo:
```js
  {
    _id: ObjectId('667d494c71cee5e4394ea810'),
    schemaVersion: 1,
    clusterId: 'mycluster',
    credential: {
      kubeConfig: '1OTOw7jb4...',
      useCspCredentials: false
    },
    resource: {
      name: 'mycluster',
      region: 'us-east-2',
      accountId: '123456789012',
      accountName: 'aws',
      vpcId: 'vpc-01b23d36a9a724c11',
      version: '1.30',
      platform: 'kops',
      public: true,
      project: null,
      compartment: null,
      networkMode: 'OVERLAY',
      tags: [ { key: 'foo', value: 'bar' } ]
    }
  }
```